app/src/main/kotlin/compiler/parser/analysis/GrammarAnalysis.kt

```diff
@@ -1,14 +1,69 @@
 package compiler.parser.analysis
 
+import compiler.common.dfa.state_dfa.DfaState
 import compiler.parser.grammar.AutomatonGrammar
+import compiler.parser.grammar.Production
 
 class GrammarAnalysis<S : Comparable<S>> {
     fun computeNullable(grammar: AutomatonGrammar<S>): Set<S> {
         return TODO()
     }
 
     fun computeFirst(grammar: AutomatonGrammar<S>, nullable: Set<S>): Map<S, Set<S>> {
-        return TODO()
+        val symbols = grammar.productions.keys.toMutableSet()
+
+        // bfs to find all grammar symbols
+        for (prod in grammar.productions) {
+            val dfa = prod.value
+            val visited = emptySet<DfaState<S, Production<S>>>().toMutableSet()
+            val queue = mutableListOf(dfa.startState)
+
+            while (queue.isNotEmpty()) {
+                val state = queue.removeAt(0)
+                if (visited.contains(state))
+                    continue
+                visited.add(state)
+                symbols.addAll(state.possibleSteps.keys)
+                queue.addAll(state.possibleSteps.values)
+            }
+        }
+
+        val first = emptyMap<S, MutableSet<S>>().toMutableMap()
+
+        // bfs through nullable edges
+        for (symbol in symbols) {
+            first[symbol] = mutableSetOf(symbol)
+            val dfa = grammar.productions[symbol]
+            if (dfa === null)
+                continue
+            val visited = emptySet<DfaState<S, Production<S>>>().toMutableSet()
+            val queue = mutableListOf(dfa.startState)
+
+            while (queue.isNotEmpty()) {
```

Comment:
Also we can try to abstract the implementation of bfs over automaton into some function with action lambda as parameter to avoid repetition.